File: src/test/java/com/gitlab/lae/intellij/actions/tree/json/ActionNodeParserTest.java
package com.gitlab.lae.intellij.actions.tree.json;

import com.gitlab.lae.intellij.actions.tree.ActionNode;
import com.gitlab.lae.intellij.actions.tree.When;
import org.junit.Test;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.List;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static javax.swing.KeyStroke.getKeyStroke;
import static org.junit.Assert.assertEquals;

public final class ActionNodeParserTest {

    @Test
    public void deserialization() throws Exception {
        List<ActionNode> expected = asList(
                ActionNode.create(
                        "ActionsTree1",
                        "Unnamed",
                        null,
                        false,
                        When.toolWindow("Project"),
                        singletonList(getKeyStroke("ctrl C")),
                        singletonList(
                                ActionNode.create(
                                        "ActionsTree2",
                                        "Unnamed",
                                        null,
                                        false,
                                        When.ALWAYS,
                                        singletonList(getKeyStroke("P")),
                                        asList(
                                                ActionNode.create(
                                                        "CloseProject",
                                                        "Unnamed",
                                                        null,
                                                        false,
                                                        When.ALWAYS,
                                                        asList(
                                                                getKeyStroke(
                                                                        "K"),
                                                                getKeyStroke(
                                                                        "ctrl K")
                                                        ),
                                                        emptyList()
                                                ),
                                                ActionNode.create(
                                                        "OpenProjectGroup",
                                                        "Unnamed",
                                                        "SEP",
                                                        false,
                                                        When.ALWAYS,
                                                        singletonList(
                                                                getKeyStroke(
                                                                        "P")),
                                                        emptyList()
                                                )
                                        )
                                )
                        )
                ),
                ActionNode.create(
                        "ActionsTree3",
                        "b",
                        null,
                        false,
                        When.fileExt("java"),
                        singletonList(getKeyStroke("ctrl X")),
                        singletonList(
                                ActionNode.create(
                                        "EditorSwapSelectionBoundaries",
                                        "Unnamed",
                                        null,
                                        true,
                                        When.ALWAYS,
                                        singletonList(getKeyStroke("ctrl X")),
                                        emptyList()
                                )
                        )
                ),
                ActionNode.create(
                        "ActionsTree4",
                        "c",
                        null,
                        false,
                        When.any(
                                When.fileExt("txt"),
                                When.all(
                                        When.toolWindow("Run"),
                                        When.fileExt("java")
                                )
                        ),
                        emptyList(),
                        emptyList()
                )
        );

        List<ActionNode> actual;
        try (InputStream stream = ActionNodeParserTest.class
                .getResourceAsStream("test.json");
             Reader reader = new InputStreamReader(stream, UTF_8)) {
            actual = ActionNodeParser.parseJsonActions(reader);
        }

        assertEquals(expected, actual);
    }

}

File: src/test/java/com/gitlab/lae/intellij/actions/tree/json/ActionNodeParserTest.kt
package com.gitlab.lae.intellij.actions.tree.json

import com.gitlab.lae.intellij.actions.tree.ActionNode
import com.gitlab.lae.intellij.actions.tree.When
import org.junit.Assert.assertEquals
import org.junit.Test
import javax.swing.KeyStroke.getKeyStroke
import kotlin.text.Charsets.UTF_8

class ActionNodeParserTest {

  @Test
  fun deserialization() {
    val expected = listOf(
      ActionNode(
        "ActionsTree.Node.1",
        "Unnamed",
        null,
        false,
        When.toolWindowActive("Project"),
        listOf(getKeyStroke("ctrl C")),
        listOf(
          ActionNode(
            "ActionsTree.Node.2",
            "Unnamed",
            null,
            false,
            When.ALWAYS,
            listOf(getKeyStroke("P")),
            listOf(
              ActionNode(
                "CloseProject",
                "Unnamed",
                null,
                false,
                When.ALWAYS,
                listOf(
                  getKeyStroke("K"),
                  getKeyStroke("ctrl K")
                ),
                emptyList()
              ),
              ActionNode(
                "OpenProjectGroup",
                "Unnamed",
                "SEP",
                false,
                When.ALWAYS,
                listOf(getKeyStroke("P")),
                emptyList()
              )
            )
          )
        )
      ),
      ActionNode(
        "ActionsTree.Node.3",
        "b",
        null,
        false,
        When.fileExtension("java"),
        listOf(getKeyStroke("ctrl X")),
        listOf(
          ActionNode(
            "EditorSwapSelectionBoundaries",
            "Unnamed",
            null,
            true,
            When.ALWAYS,
            listOf(getKeyStroke("ctrl X")),
            emptyList()
          )
        )
      ),
      ActionNode(
        "ActionsTree.Node.4",
        "c",
        null,
        false,
        When.any(
          When.fileExtension("txt"),
          When.all(
            When.toolWindowActive("Run"),
            When.fileExtension("java")
          )
        ),
        emptyList(),
        emptyList()
      )
    )

    val actual = ActionNodeParserTest::class.java
      .getResourceAsStream("test.json").use {
        ActionNodeParser.parseJsonActions(it.reader(UTF_8))
      }

    assertEquals(expected, actual)
  }
}
